docs/design/kuadrant-api-v2.md

```diff
@@ -0,0 +1,178 @@
+* A (base) URL
+* A kubernetes service reference (name and namespace).
+
+When the source is a kubernetes service, kuadrant provides annotations and labels that can be added
+to the service. These annotations and labels can specify, for example, the OAS reference.
+These annotations and labels aim to make the `spec` of the *API Product* resource simpler to specify.
+
+**AuthConfig Templates**
+
+The configuration of the API authentication requires *three* steps. Only the first
+two steps involve kuadrant, but all three of them are required to have the system working.
+For completeness, all of them are listed here.
+
+1. Operation based authentication config
+
+The customer needs to specify which API operation (method + path) requires authentication and
+which type of authentication (API key, oauth2, oidc, http basic, etc). OpenAPI spec fully covers
+this use case and kuadrant can read this configuration from OAS doc. For any authN type not covered
+by OAS 3.X, kuadrant could provide extensions for added authentication types. OpenAPI 3.x does
+not support authentication requirements at path level, only global or operation level.
+
+1. Configuration for each type of authentication
+
+The customer needs to specify the authentication provider configuration. Kuadrant provides Authorino
+`AuthConfig` custom resource to specify authentication provider configuration. The *API Product*
+resource will include a template object in the `.spec` to specify `AuthConfig` objects.
+
+The *API Product* resource allows to define multiple AuthConfig templates.
+
+To illustrate this with examples, the API Key auth type would require `AuthConfig` template with
+`spec.identity[].apiKey` object. The OIDC auth type would require `AuthConfig` template with
+`spec.identity[].oidc` object.
+
+Kuadrant can add some validations to prevent inconcistencies. For instance, if the OAS doc specifies
+that API key is being required for some API operations, then kuadrant can make sure that the
+`AuthConfig` template includes `apiKey` identity type.
+
+Actually, Authorino's `AuthConfig` covers far more than authentication. The customer can do more
+than just authentication, the customer can apply authorization policies as well.
+
+1. Authentication data provisioning
+
+The customer needs to provision some authentication data. For the API key use case, the customer
+needs to generate secrets with the actual tokens. For the OIDC use case, the customer needs to
+provision OAUTH2 clients and users in the Authentication provider.
+
+**RateLimit Template**
+
+The configuration of the API rate limit requires *two* steps.
+
+1. Path/Operation based rate limit config
+
+The customer needs to define which API paths/operations require rate limit enforcing.
+Furthermore, the customer needs to define what type of rate limit needs to be enforced at
+path/operation level. Global rate limit is a particular case where all API operations require
+enforcing rate limit of a particular rate limit type.
```

Comment:
for rate limiting I think we need to allow for headers also

Reply:
It could be. But then, OAS cannot be used to specify *where* to apply rate limit. Instead, kuadrant needs to implement a rule matching system for rate limit. It would be similar to the [Rules of istio's Authorization Policy](https://istio.io/latest/docs/reference/config/security/authorization-policy/#Rule)

For the first implementation, I would stick to OAS. But definitely, that can be a real user requirement.